diff --git a/components/esp_lvgl_port/esp_lvgl_port.c b/components/esp_lvgl_port/esp_lvgl_port.c
index 85c47b2..e760a8c 100644
--- a/components/esp_lvgl_port/esp_lvgl_port.c
+++ b/components/esp_lvgl_port/esp_lvgl_port.c
@@ -396,7 +396,7 @@ esp_err_t lvgl_port_remove_touch(lv_indev_t *touch)
     lvgl_port_touch_ctx_t *touch_ctx = (lvgl_port_touch_ctx_t *)indev_drv->user_data;
 
     /* Remove input device driver */
-    lv_indev_delete(touch);
+    //lv_indev_delete(touch);
 
     if (touch_ctx) {
         free(touch_ctx);
@@ -562,7 +562,7 @@ esp_err_t lvgl_port_remove_navigation_buttons(lv_indev_t *buttons)
     lvgl_port_nav_btns_ctx_t *buttons_ctx = (lvgl_port_nav_btns_ctx_t *)indev_drv->user_data;
 
     /* Remove input device driver */
-    lv_indev_delete(buttons);
+    //lv_indev_delete(buttons);
 
     if (buttons_ctx) {
         free(buttons_ctx);
@@ -642,7 +642,7 @@ esp_err_t lvgl_port_remove_usb_hid_input(lv_indev_t *hid)
     lvgl_port_usb_hid_ctx_t *hid_ctx = (lvgl_port_usb_hid_ctx_t *)indev_drv->user_data;
 
     /* Remove input device driver */
-    lv_indev_delete(hid);
+    //lv_indev_delete(hid);
 
     /* If all hid input devices are removed, stop task and clean all */
     if (lvgl_port_ctx.hid_ctx.mouse.drv.disp == NULL && lvgl_port_ctx.hid_ctx.kb.drv.disp) {
